go/porcelain/deploy_test.go

```diff
@@ -317,9 +317,38 @@ func TestUploadFiles_Errors(t *testing.T) {
 	for _, bundle := range files.Files {
 		d.Required = append(d.Required, bundle.Sum)
 	}
-	err = client.uploadFiles(ctx, d, files, nil, fileUpload, time.Minute)
+	err = client.uploadFiles(ctx, d, files, nil, fileUpload, time.Minute, false)
 	require.Equal(t, err.Error(), "[PUT /deploys/{deploy_id}/files/{path}][500] uploadDeployFile default  &{Code:0 Message:}")
 }
+func TestUploadFiles400Errors(t *testing.T) {
+	ctx := gocontext.Background()
+
+	server := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, _ *http.Request) {
+		rw.WriteHeader(http.StatusUnprocessableEntity)
```

Comment:
This is super helpful and definitely makes sense! -- looking at other test cases I see this is done similarly. However, when we set ```rw.Write([]byte(`{"message": "Unprocessable Entity", "code": 422}`))``` we keep getting the error`&{0 } (*models.Error) is not supported by the TextConsumer, can be resolved by supporting TextUnmarshaler interface`.

So then the apiError isn't set and then our test skips over the code we're trying to test and keeps retrying 🙈

Looking at other examples in other tests, it seems like it works to set the response this way when calling `client.Operations.UploadDeployFile`, but not when calling `client.uploadFiles` -- when calling `uploadFiles` the error is thrown from `UploadDeployFile`.

We've been trying to debug this by writing the body different ways, trying to marshal the response etc, but haven't had any success yet 🤔 Is there maybe something silly we're missing?

Reply:
This is where the error is being thrown from: https://github.com/netlify/open-api/blob/master/go/plumbing/operations/operations_client.go#L4131

Reply:
I think It might be the order on which those are set, mockserver is a bit quirky at times. try if setting the headers in this order helps:

```go
rw.Header().Set("Content-Type", "application/json; charset=utf-8")
rw.WriteHeader(http.StatusUnprocessableEntity)
rw.Write([]byte(`{"message": "Unprocessable Entity", "code": 422 }`))
```

Reply:
👋 @4xposed thanks so much for the tip! That worked and now as far as I can tell, the server is responding as expected. However, now we're hitting a different problem and struggling to debug it. 

We now receive the error with the correct code and message, but when we get to [this line](https://github.com/netlify/open-api/pull/478/files#diff-602e1ebf353d4906b9de529630317e7f88f5832aa50bbb1ca269741412e1c112R542)
```apiErr, ok := operationError.(apierrors.Error)```
no matter what we try, apiErr is always `nil` and `ok` is always false. 

`operationError` is an `error` that looks like:
```error(*github.com/netlify/open-api/v2/go/plumbing/operations.UploadDeployFunctionDefault) *{_statusCode: 422, Payload: *github.com/netlify/open-api/v2/go/models.Error {Code: 422, Message: "Unprocessable Entity"}}```

In my debugger, I can access the status code by calling either `operationError._statusCode` or `operationError.Payload.Code`, but not by calling `operationError.(apierrors.Error)`. However, I can't call `operationError.Payload.Code` in the actual code, because `Payload`/`Code` isn't defined on the Error type. 

So since `operationError.(apierrors.Error)` never seems to evaluate to anything and there wasn't a test for the original code, we're uncertain if we're doing something wrong here, or if the original code was actually working as intended. 

Are we missing something here? Is there a better way to test this (perhaps non-locally?) 

cc @jrwhitmer @JGAntunes

Reply:
Hey @jenae-janzen, I have been looking into this and I think this casting to `apierrors.Error` never actually worked 😅 

I opened this PR to fix that issue: https://github.com/netlify/open-api/pull/486

once that's merged you should be able to rebase and resume your work on this PR

Reply:
@jenae-janzen this has been merged - https://github.com/netlify/open-api/pull/486 - I think if we rebase the errors should now behave as expected 👍

>  or if the original code was actually working as intended.

Spoiler alert, it never worked as intended 😂

Reply:
@jenae-janzenI believe you will be in Europe soon-ish, if you want to pair on this feel free to book something in my calendar!

Reply:
Pairing might not be necessary since most of this seems to be working, but I'd appreciate you taking another look when you have the time 🙂

Reply:
we could call `err.Unwrap()` seems like `backoff.PermanentError` has a few methods for this: https://github.com/cenkalti/backoff/blob/v4/retry.go#L129

I don't think it's strictly necessary, but it would be nice to at least have some way that we have the _right_ error and not _any_ error, if that makes sense.

What do you think?

Reply:
I'll give it a shot! I do agree it would be nice to have the right error if possible

Reply:
Okay, I used `Require. ErrorContains` instead and that seemed to do the trick for verifying we have the correct error message